crates/sui-graphql-rpc/src/types/cursor.rs

```diff
@@ -0,0 +1,357 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{fmt, ops::Deref};
+
+use async_graphql::{
+    connection::{CursorType, OpaqueCursor},
+    *,
+};
+use serde::{de::DeserializeOwned, Serialize};
+
+use crate::{config::ServiceConfig, error::Error};
+
+/// Wrap the `OpaqueCursor` type to implement Scalar for it.
+pub(crate) struct Cursor<C>(OpaqueCursor<C>);
+
+/// Connection field parameters parsed into a single type that encodes
+#[derive(Debug)]
+pub(crate) struct Page<C> {
+    after: Option<Cursor<C>>,
+    before: Option<Cursor<C>>,
+    limit: u64,
+    side: Side,
+}
+
+/// Whether the page is extracted from the beginning or the end of the range bounded by the cursors.
+#[derive(PartialEq, Eq, Debug)]
+enum Side {
+    Prefix,
+    Suffix,
+}
+
+impl<C> Cursor<C> {
+    pub(crate) fn new(cursor: C) -> Self {
+        Cursor(OpaqueCursor(cursor))
+    }
+}
+
+impl<C> Page<C> {
+    pub(crate) fn from_params(
```

Comment:
This function takes over the job of `validate_cursor_pagination`.  By taking all the parameters by value, and returning a type, it makes sure that we can't misuse connection parameters (forget to check for max, or set default, etc), as long as we use this everywhere -- which we can make more likely by using it as the accepted type in the data layer.